pandas/core/reshape/util.py

```diff
@@ -40,7 +40,13 @@ def cartesian_product(X):
     cumprodX = np.cumproduct(lenX)
 
     if np.any(cumprodX < 0):
-        raise ValueError("Product space too large to allocate arrays!")
+        raise ValueError(
+            "Product space too large to allocate arrays! If calling "
+            "`DataFrame.groupby`, you are trying to group over too complex "
```

Comment:
Instead of changing the error message here, I think it's better if you catch this error wherever it's raising in the groupby code and reraise your new message

```
# in some groupby code

try:
    ....
except ValueError as err:
    # check for this original message
    raise ValueError("You are ...") from err
```